lua/leap-by-word.lua

```diff
@@ -15,7 +15,55 @@ local function get_input()
   end
 end
 
-local function get_targets(winid, char, direction)
+-- patterns and functions for testing if a character should be considered a target
+local matches = {
+  upper = [=[\v[[:upper:]]\C]=],
+  lower = [=[\v[[:lower:]]\C]=],
+  digit = [=[\v[[:digit:]]\C]=],
+  word  = [=[\v[[:upper:][:lower:][:digit:]]\C]=],
+  keyw  = [=[\v[[:keyword:]]\C]=],
+}
+local function test(char, match)
+  if char == nil then return false -- vim.fn.match returns false for nil char, but not if pattern contains `[:lower:]`
+  else return vim.fn.match(char, match) == 0 end
+end
+
+local function test_split_identifiers(chars, cur_i)
+  local cur_char = chars[cur_i]
+
+  local is_match = false
+
+  if test(cur_char, matches.upper) then
+    local prev_char = chars[cur_i - 1]
+    if not test(prev_char, matches.upper) then is_match = true
+    else
+      local next_char = chars[cur_i + 1]
+      is_match = test(next_char, matches.word) and not test(next_char, matches.upper)
+    end
+  elseif test(cur_char, matches.digit) then
+    is_match = not test(chars[cur_i - 1], matches.digit)
+  elseif test(cur_char, matches.lower) then
+    is_match = not test(chars[cur_i - 1], matches.word) or test(chars[cur_i - 1], matches.digit)
+  else
+    local prev_char = chars[cur_i - 1]
+    is_match = prev_char ~= cur_char -- matching only first character in ==, [[ and ]]
+  end
+
+  return is_match
+end
+local function test_regular(chars, cur_i)
+  local cur_char = chars[cur_i]
+  local prev_char = chars[cur_i - 1]
+  if test(cur_char, matches.keyw) then return not test(prev_char, matches.keyw)
+  else return prev_char ~= cur_char end
+end
+
+local function get_targets(winid, char, direction, test_func)
+  -- Case sensitive pattern:
+  -- local match_regex = "\\v[[="..char.."=]]\\C"
+  -- Ignore case pattern:
+  local match_regex = "\\v[[="..char.."=]]\\c" -- no escaping is fine ?  -- doesn't work: '\\v(%d'..tostring(vim.fn.char2nr(char))..')\\c'
```

Comment:
what do you mean by these comments?

Reply:
I wasn't sure if the character in that position needs escaping